sys-apps/busybox: fix some bugs

ceamac opened the pull request:
I made some bugfixes.

1.34.1 is stable, I made all changes in 1.34.1-r1.

Remaining bugs:
- https://bugs.gentoo.org/824222 - this requires only clean-up.  I deleted 3 older versions but the diff on github looks very chaotic, so I reverted it.  They can be removed later in a different PR.
- https://bugs.gentoo.org/836920 - security bug.  I looked upstream and nothing.  There is one question about this on their mailing list and no answer.  If upstream does not fix them properly we can apply the patches from alpine, but I think we should wait for upstream for a reasonable amount of time.

ceamac commented on sys-apps/busybox/busybox-1.34.1-r1.ebuild:
```diff
@@ -0,0 +1,358 @@
+		use make-symlinks || dosym /bin/bb /sbin/mdev
+		cp "${S}"/examples/mdev_fat.conf "${ED}"/etc/mdev.conf || die
+		if [[ ! "$(get_libdir)" == "lib" ]]; then
+			sed -i -e "s:/lib/:/$(get_libdir)/:g" "${ED}"/etc/mdev.conf || die
```
https://bugs.gentoo.org/831251 - replace `lib` with `lib64` where appropriate

ceamac commented on sys-apps/busybox/busybox-1.34.1-r1.ebuild:
```diff
@@ -0,0 +1,358 @@
+	dodoc inittab depmod.pl *.conf *.script undeb unrpm
+
+	cd ../networking || die
+	dodoc httpd_indexcgi.c httpd_post_upload.cgi
```
https://bugs.gentoo.org/828467 - copy CGI scripts in examples

ceamac commented on sys-apps/busybox/busybox-1.34.1-r1.ebuild:
```diff
@@ -0,0 +1,358 @@
+	# for compatibility, provide /usr/bin/env
+	mkdir -p _install/usr/bin || die
+	if [[ ! -e _install/usr/bin/env ]]; then
+		ln -s /bin/env _install/usr/bin/env || die
```
https://bugs.gentoo.org/825718 - the source already exists on some systems, simply skip it

ceamac commented on sys-apps/busybox/busybox-1.34.1-r1.ebuild:
```diff
@@ -0,0 +1,358 @@
+		newinitd "${FILESDIR}"/watchdog.initd busybox-watchdog
+	fi
+	if busybox_config_enabled UDHCPC; then
+		sed -i 's:$((metric++)):$metric; metric=$((metric + 1)):' examples/udhcp/simple.script || die
```
https://bugs.gentoo.org/801535 - I put it all on one line to be easier to use sed.  I'll also leave a comment in the bug about the other concerns.

ceamac commented on sys-apps/busybox/busybox-1.34.1-r1.ebuild:
```diff
@@ -0,0 +1,358 @@
+		doins examples/udhcp/udhcpd.conf
+	fi
+	if busybox_config_enabled ASH && ! use make-symlinks; then
+		dosym -r /bin/busybox /bin/ash
```
https://bugs.gentoo.org/738542 - I don't think we really need a USE flag for this - simply symlink it.  I am not aware of any `/bin/ash` to collide to.

ceamac commented on sys-apps/busybox/busybox-1.34.1-r1.ebuild:
```diff
@@ -0,0 +1,358 @@
+	busybox_config_option n MONOTONIC_SYSCALL
+	busybox_config_option n USE_PORTABLE_CODE
+	busybox_config_option n WERROR
+	busybox_config_option n MODPROBE_SMALL #472464
```
https://bugs.gentoo.org/472464 - `CONFIG_MODPROBE_SMALL=y` disables `depmod.c` and uses a smaller one that does not support `-b`.  Setting this to no creates slightly larger and slightly more useful modutils.

ceamac commented on sys-apps/busybox/busybox-1.34.1-r1.ebuild:
```diff
@@ -0,0 +1,358 @@
+	fi
+
+	# setup the config file
+	emake -j1 -s allyesconfig ENABLE_SELINUX=$(usex selinux y n) >/dev/null #620918
```
https://bugs.gentoo.org/620918 - This ebuild first creates a config with all settings set to `yes`, then starts to change them to `no` selectively.  The problem here was the Makefile searches for libselinux using pkg-config and emits that warning (which is harmless).  If I run `emake` with `ENABLE_SELINUX=n` it no longer searches for selinux, even if it sets it to `y`.  I think I could have hardcoded it to `n` here, because it runs again later, but I don't have a selinux profile to test this, so this looks safer.

ceamac replied:
Didn't work, I had to patch the Makefile.

ceamac commented on sys-apps/busybox/files/crond.initd:
```diff
@@ -0,0 +1,12 @@
+depend() {
+	need clock logger
+	provide cron
+}
```
Both files copied from cronie and edited, so I kept the copyright and I updated the year.

ceamac commented on sys-apps/busybox/busybox-1.34.1-r1.ebuild:
```diff
@@ -0,0 +1,363 @@
+
+	# setup the config file
+	emake -j1 -s allyesconfig SKIP_SELINUX=$(usex selinux n y) >/dev/null #620918
+	# nommu forces a bunch of things off which we want on #387555
```
setting SKIP_SELINUX skips searching for selinux at this stage.  We don't need to search now, this is an allyesconfig that will be changed later.

ceamac commented on sys-apps/busybox/busybox-1.34.1-r1.ebuild:
```diff
@@ -0,0 +1,363 @@
+	# nommu forces a bunch of things off which we want on #387555
+	busybox_config_option n NOMMU
+	sed -i '/^#/d' .config
+	yes "" | emake -j1 -s oldconfig SKIP_SELINUX=$(usex selinux n y) >/dev/null #620918
```
Same here.  The warnings were harmelss, but they still appeared.

ceamac commented on sys-apps/busybox/busybox-1.34.1-r1.ebuild:
```diff
@@ -0,0 +1,363 @@
+		busybox_config_option n ${opt}
+	done
+
+	emake -j1 oldconfig > /dev/null
```
This is the final make that searches for selinux if still enabled.

thesamesam commented:
This looks good to me, but I'd ask that you put your comments inside the ebuild too where they're explanatory -- it'll help future travellers!

ceamac commented:
I added comments where it's not obvious why I made a change.

Also for the last commit I fixed almost all the problems reported by `shellcheck` but I cannot test it as I don't have a system with `busybox` and `mdev` atm.  So I'm leaving this as draft until I can test it.

One report remaining
```
In ide_links line 20:
num=$(ls "${media}"[0-9]* 2>/dev/null | wc -l)
      ^-----------------------------^ SC2012 (info): Use find instead of ls to better handle non-alphanumeric filenames.
```

From my point of view replacing `ls` with `find` in this instance doesn't add any benefit.

thesamesam commented:
Really, fantastic work here. This is excellent!

thesamesam commented:
As a followup, @ceamac, could you sync live & 1.35.0? (revbump that too, even though unkw'd)
